packages/typescript-estree/src/parseSettings/getProjectConfigFiles.ts

```diff
@@ -0,0 +1,63 @@
+import debug from 'debug';
+import * as fs from 'fs';
+import * as path from 'path';
+
+import type { ParseSettings } from '.';
+
+const log = debug('typescript-eslint:typescript-estree:getProjectConfigFiles');
+
+const tsconfigMatchCache = new Map<string, string | undefined>();
+
+/**
+ * @remarks Only use this for tests!
+ */
+export function clearMatchCacheForTests(): void {
+  tsconfigMatchCache.clear();
+}
+
+/**
+ * Checks for a matching TSConfig to a file including its parent directories,
+ * permanently caching results under each directory it checks.
+ *
+ * @remarks
+ * We don't (yet!) have a way to attach file watchers on disk, but still need to
+ * cache file checks for rapid subsequent calls to fs.existsSync. See discussion
+ * in https://github.com/typescript-eslint/typescript-eslint/issues/101.
+ */
+export function getProjectConfigFiles(
+  parseSettings: Pick<ParseSettings, 'filePath' | 'tsconfigRootDir'>,
+  project: string | string[] | true | undefined,
+): string | string[] | undefined {
+  if (project !== true) {
+    return project;
+  }
+
+  log('Looking for tsconfig.json at or above file: %s', parseSettings.filePath);
+  let directory = path.dirname(parseSettings.filePath);
+  const checkedDirectories = [directory];
+
+  do {
+    log('Checking tsconfig.json path: %s', directory);
+    const tsconfigPath = path.join(directory, 'tsconfig.json');
+    const cached =
+      tsconfigMatchCache.get(directory) ??
```

Comment:
the only thing we need to be weary of is that this cache can get out-of-sync with the real state on disk for "persistent" runs.

we should provide a cache lifetime and nuke the cache at the start of this function if it's expired.

[`eslint-plugin-import` uses a custom Map class](https://github.com/import-js/eslint-plugin-import/blob/766af5ff6c2e97cbe2a2a211729f7c129da3005d/utils/ModuleCache.js#L6) which doesn't return cache entries older than the configured lifetime.
[They have a default cache lifetime of 30s](https://github.com/import-js/eslint-plugin-import/blob/766af5ff6c2e97cbe2a2a211729f7c129da3005d/utils/ModuleCache.js#L35-L37), so maybe we could use that as well? Could be a parser option if we wanted to allow people to reduce or increase it if they want.

(I got carried away and wrote up what I mean sorry for stealing the problem)
```ts
const LIFETIME_SECONDS: number | 'Infinity' = 30;

class TSConfigCache {
  #map = new Map<
    string,
    Readonly<{
      value: string | undefined,
      lastSeen: readonly [number, number],
    }>
  >();
  #reverseMap = new Map<string, string[]>();

  set(key, value): void {
    this.map.set(key, { value, lastSeen: process.hrtime() });
    const reverseEntry = this.reverseMap.get(key) ?? [];
    reverseEntry.push(value);
    this.reverseMap.set(key, reverseEntry);
  }

  get(key): string | undefined {
    const entry = this.map.get(key);
    if (entry) {
      if (LIFETIME_SECONDS === 'Infinity' || process.hrtime(entry.lastSeen)[0] < LIFETIME_SECONDS) {
        // cache hit woo!
        return entry.value;
      } else {
        // key has expired - clean it up to free up memory
        const reverseEntry = this.reverseMap.get(entry.value);
        if (reverseEntry) {
          for (const expiredKey of reverseEntry) {
            this.map.delete(expiredKey);
          }
          this.reverseMap.delete(entry.value);
        }
      }
    }
    // no hit :'(
    return undefined;
  }
}
```

Also worth noting - we could use this same cache class to cache our glob lookups!
Currently we don't cache glob lookups and so they're a point of slowness.

Also worth noting - if we're in single-run mode we can actually set `LIFETIME_SECONDS = 'Infinity'` because we treat the entire disk as constant when we're single-running! Which should save us a lot of time - esp for cases of `**` globs.

Reply:
For reference - I implemented the cache classes in my pr here https://github.com/typescript-eslint/typescript-eslint/pull/6367

Reply:
Nice, I reused the class there. Thanks!